[#9] 전역적 예외 처리를 위한 예외 advice 기능 추가

daakludens opened the pull request:
프로젝트 내에서 일어날 수 있는 모든 예외 경우를 처리하기 위한 범용적 예외 처리 기능 추가

daakludens committed:
[#9] 전체 예외 경우를 처리하기 위한 ExceptionAdvice 클래스 및 예외 에러 코드와 메시지를 출력하기 위한 클래스 생성

f-lab-dev commented on src/main/java/com/ludensdomain/advice/ExceptionAdvice.java:
```diff
@@ -0,0 +1,22 @@
+    @ExceptionHandler(UnauthorizedUserException.class)
+    @ResponseStatus(HttpStatus.UNAUTHORIZED)
+    protected ExceptionResponse unauthorizedUserException() {
+        return exceptionService.getResult(-1001, "해당 기능을 이용할 수 있는 권한이 없습니다.");
```
이 `-1001`은 어떤 것을 뜻하나요? 전달해주는 의미가 있을까요?

daakludens replied:
메시지는 유저를 위한 기능이고 -1001의 경우는 개발자를 위한 기능이라고 생각했습니다. 코드로 예외를 분류하고 예외 발생 시 코드로 다른 일을 할 수 있지 않을까 생각해서 넣었습니다. 
메시지 하나만으로도 충분한가요?

f-lab-dev replied:
클라이언트에서 권한이 없을 경우는 어떻게 처리해야겠다라는 시나리오가 나온게 아니라면 굳이 미리 생각해서 넣어줄 필요는 없을 것 같습니다~

daakludens replied:
네, 알겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/advice/ExceptionResponse.java:
```diff
@@ -0,0 +1,14 @@
+import lombok.Setter;
+
+@Getter
+@Setter
```
이 클래스는 setter가 필요한가요?

daakludens replied:
ExceptionAdvice에서 받는 String message를 넣어주기 위해 Setter가 필요하다고 생각했습니다. 
Setter 사용을 지양하기 위해 메시지를 넣어줄 다른 방식을 찾아보겠습니다.

f-lab-dev replied:
네 꼭 필요하다면 넣어줘야겠지만 immutable을 지켜줌으로써 오는 장점이 정말 많습니다. 이것도 계속 고민해보시면 좋을 것 같아요~

f-lab-dev commented on src/main/java/com/ludensdomain/advice/exceptions/ExceptionService.java:
```diff
@@ -0,0 +1,22 @@
+     * @param message 에러 메시지
+     * @return response 에러 코드와 메시지를 담은 ExceptionResponse
+     */
+    public ExceptionResponse getResult(int code, String message) {
```
이거는 비즈니스로직보다는 유틸 메소드에 더 가까운 것 같습니다. 그럼 어떻게 바꿔야할지 한번 고민해보시면 좋을 것 같네요~

daakludens replied:
유틸 메서드라면 ExceptionAdvice에 하나의 메서드 형태로 넣어주면 될 것 같습니다. 
유틸 메서드는 반복적으로 자주 사용되는 메서드를 의미하는데, 이 경우 모든 예외 처리를 실행할 메서드이기에 적용할 수 있습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/controller/GameController.java:
```diff
@@ -19,6 +21,8 @@
 @RequiredArgsConstructor
 public class GameController {
 
+    final Logger logger = LoggerFactory.getLogger(GameController.class);
```
1. 접근제한자도 붙여주세요~
2. lombok으로도 대체가 가능합니다~

daakludens replied:
1. 접근제한자는 private를 붙일 수 있습니다.
2. lombok 사용 시 `@Log4j2`를 사용할 수 있습니다.

두 방식은 똑같은 라인을 생성합니다. lombok이 설치되어 있다면 간편하게 @Log4j2를 사용하는 게 나은 것 같습니다.

f-lab-dev reviewed (commented):
PR 제목이 너무 범위가 큰것 같습니다. 이 작업 내용을 더 명확하게 보여줄 수 있는 제목을 지어주는게 좋을 것 같아요~

daakludens committed:
[#9] refactor:
1. ExceptionService 내 메서드를 유틸 메서드로써 ExceptionAdvice에 추가
2. GameController에 로그 방식 변경
3. ExceptionResponse에 불필요한 코드 숫자와 Getter 메서드 삭제

f-lab-dev commented on src/main/java/com/ludensdomain/advice/ExceptionResponse.java:
```diff
@@ -1,14 +1,10 @@
-import lombok.Getter;
 import lombok.Setter;
 
-@Getter
```
여기는 잘못 빼신 것 같습니다~

daakludens replied:
네, 수정했습니다.

daakludens commented:
PR 제목 수정했습니다.

daakludens committed:
[#9] refactor : ExceptionResponse에 @Setter 대신 @Builder 추가

daakludens committed:
[#9] refactor : getResult 메서드에 builder 적용
